pkg/restic/repository_ensurer.go

```diff
@@ -137,3 +169,30 @@ func (r *repositoryEnsurer) getReadyChan(name string) chan *velerov1api.ResticRe
 	r.readyChans[name] = make(chan *velerov1api.ResticRepository)
 	return r.readyChans[name]
 }
+
+func (r *repositoryEnsurer) lock(volumeNamespace, backupLocation string) {
+	r.repoLocksLock.Lock()
+	defer r.repoLocksLock.Unlock()
+
+	repoKey := repoKey{
+		volumeNamespace: volumeNamespace,
+		backupLocation:  backupLocation,
+	}
+
+	if _, ok := r.repoLocks[repoKey]; !ok {
+		r.repoLocks[repoKey] = new(sync.Mutex)
+	}
+
+	r.repoLocks[repoKey].Lock()
+}
+
+func (r *repositoryEnsurer) unlock(volumeNamespace, backupLocation string) {
+	repoKey := repoKey{
+		volumeNamespace: volumeNamespace,
+		backupLocation:  backupLocation,
+	}
+
+	r.repoLocks[repoKey].Unlock()
```

Comment:
Does this need to be wrapped by `r.repoLocksLock`, since it's technically a mutation?

Reply:
yeah, it does need a read lock...i need to do a little re-jiggering to avoid deadlocks.

Reply:
Is it a readlock to call the underlying `Lock` mechanism? I'd think it would be a write, since state in the key's mutex is changing.

Reply:
OK, the latest commit should clean this up.  `repoLocksLock` (now `repoLocksMu`) synchronizes writes/reads to the `repoLocks` map, since they're not threadsafe. And each mutex in the map synchronizes runs of `EnsureRepo` for a given namespace + location pair.  The previous method structure had interleaved mutexes which was incorrect because it leads to deadlocks.  That's no longer the case.

Reply:
I'll retest if there's a new image, @skriss?  Let me know the tag to pull.

Reply:
@PeterGrace I pushed `steveheptio/ark:v0.10.2-alpha.2` with the latest code